### PR TITLE
SDN-5234: rebase k8s v1.31.1

### PR DESCRIPTION
**Pull request by jluhrsen**
[SDN-5234](https://issues.redhat.com//browse/SDN-5234): rebase k8s v1.31.1
    
    Update with:
    
    go get -u github.com/openshift/build-machinery-go
    go get -u k8s.io/api
    go get -u k8s.io/apimachinery
    go get -u k8s.io/code-generator
    go get -u k8s.io/component-base
    go get -u k8s.io/klog/v2
    go get -u k8s.io/kube-proxy
    go get -u k8s.io/utils
    go get -u sigs.k8s.io/controller-runtime
    go get -u github.com/openshift/api
    go get -u github.com/openshift/client-go
    go get -u github.com/openshift/library-go
    go get -u github.com/openshift/machine-config-operator
    go get -u k8s.io/apiextensions-apiserver
    go get -u k8s.io/client-go
    go get -u sigs.k8s.io/controller-tools
    go mod tidy
    go mod vendor
    make verify
    
    Build was breaking due to a change in openshift/library-go [0].
    Implemented some new functions for:
    
    genericoperatorclient.NewClusterScopedOperatorClient()
    v1helpers.OperatorClient
    
    [0] https://github.com/openshift/library-go/pull/1787

### CODE REVIEW

**Comment by jluhrsen** [comment]
/hold

this has to wait until the openshift/library-go rebase is done and this PR is updated with that. we need [this commit](https://github.com/openshift/library-go/pull/1771/commits/078920fc4e89f583691047e9904fd174d5796bbd).
It was broken with [this apimachinery change](https://github.com/kubernetes/apimachinery/commit/126f5cee563da8f0b090df60ca8738d16b5188bf).

**Comment by atiratree** [comment]
hi @jluhrsen https://github.com/openshift/api/commit/c40ad4a09eda1eb8801a3e4de805d25f19ac537e change has been merged to openshift/api.

CNO has to react to these changes, can you please pick the changes from https://github.com/openshift/cluster-network-operator/pull/2506 into your PR?

**Comment by jluhrsen** [comment]
> hi @jluhrsen [openshift/api@c40ad4a](https://github.com/openshift/api/commit/c40ad4a09eda1eb8801a3e4de805d25f19ac537e) change has been merged to openshift/api.
> 
> CNO has to react to these changes, can you please pick the changes from #2506 into your PR?

thank you @atiratree . Yes, I will take care of this.

File: pkg/client/operatorclient.go [hunk 1]
```diff
@@ -21,6 +22,17 @@ type OperatorHelperClient struct {
 	client   operatorclientv1.NetworkInterface
 }
 
+// ApplyOperatorSpec and ApplyOperatorStatus were added here:
+// https://github.com/openshift/library-go/commit/2e06abc0a54d9ecd0a8783325068896c347bd137#diff-0e7fb0f95ea48e43d0cad6743df9338e45348ebb2ca9aa0c5491eda440642b45
+// added as no-ops for now.
+// TO-DO: investigate if it's better to implement and use these.
```
Comment (danwinship) [inline pkg/client/operatorclient.go:28]: I think they get called from the library code even if you don't call them directly...

Comment (jluhrsen) [reply pkg/client/operatorclient.go:28]: @danwinship , are you saying I should find a way to implement these better then?

**Comment by atiratree** [comment]
@jluhrsen btw it seems you are still missing this commit: https://github.com/openshift/cluster-network-operator/pull/2506/commits/febd41376cc3553fa47a3be8cb9503337a80934d

**Comment by jluhrsen** [comment]
> @jluhrsen btw it seems you are still missing this commit: [febd413](https://github.com/openshift/cluster-network-operator/commit/febd41376cc3553fa47a3be8cb9503337a80934d)

😮 THANK YOU so much for this @atiratree, I was struggling to understand what was happening.

My last update ran CI and the e2e's passed that were checking network-operator status so looks like that's what I needed. I had to do one last update to fix the linter.

If you have any time could you give a quick review on what I added. I literally just took the pattern from a [commit](https://github.com/openshift/cluster-openshift-controller-manager-operator/pull/368/commits/7485f9280da9acd8cb40c74ff62d11db2b3a3c18)
in the cluster-openshift-controller-manager-operator project

/cc @danwinship

**Review by atiratree — COMMENTED** [review]
> If you have any time could you give a quick review on what I added. I literally just took the pattern from a commit
in the cluster-openshift-controller-manager-operator project

I am not the primary contributor/reviewer but the `ApplyOperatorSpec`/`ApplyOperatorStatus` functions look ok to me.

File: pkg/controller/connectivitycheck/connectivity_check_controller.go [hunk 2]
```diff
@@ -492,3 +501,10 @@ func Start(ctx context.Context, kubeConfig *rest.Config) error {
 
 	return nil
 }
+
+func extractOperatorSpec(obj *unstructured.Unstructured, fieldManager string) (*applyoperatorv1.OperatorSpecApplyConfiguration, error) {
+	return applyoperatorv1.OperatorSpec(), nil
+}
+func extractOperatorStatus(obj *unstructured.Unstructured, fieldManager string) (*applyoperatorv1.OperatorStatusApplyConfiguration, error) {
+	return applyoperatorv1.OperatorStatus(), nil
+}
```
[the remark this answers is not in the document]
Comment (jluhrsen) [reply pkg/controller/connectivitycheck/connectivity_check_controller.go]: update to use `nil` in `NewClusterScopedOperatorClient`

File: pkg/client/operatorclient.go [hunk 3]
```diff
@@ -80,10 +87,124 @@ func (c *OperatorHelperClient) UpdateOperatorStatus(ctx context.Context, resourc
...
+		return fmt.Errorf("unable to get operator configuration: %w", err)
+	default:
+		original, err := v1.ExtractNetwork(instance, fieldManager)
+		if err != nil {
+			return fmt.Errorf("unable to extract network operator configuration: %w", err)
+		}
+		if equality.Semantic.DeepEqual(original, desired) {
+			return nil
+		}
+	}
```
[the remark this answers is not in the document]
Comment (jluhrsen) [reply pkg/client/operatorclient.go]: ok, thank you for the help. I've updated to remove the comparison.

File: pkg/client/operatorclient.go [hunk 4]
```diff
@@ -80,10 +87,124 @@ func (c *OperatorHelperClient) UpdateOperatorStatus(ctx context.Context, resourc
...
+		FieldManager: fieldManager,
+	})
+	if err != nil {
+		return fmt.Errorf("unable to Apply network operator using fieldManager %q: %w", fieldManager, err)
+	}
+
+	return nil
+}
+
+func (c *OperatorHelperClient) ApplyOperatorStatus(ctx context.Context, fieldManager string, desiredStatus *v1.OperatorStatusApplyConfiguration) error {
```
[the remark this answers is not in the document]
Comment (jluhrsen) [reply pkg/client/operatorclient.go:118]: removed the comparison and kept the timing stuff out of it.

File: pkg/client/operatorclient.go [hunk 5]
```diff
@@ -80,10 +88,129 @@ func (c *OperatorHelperClient) UpdateOperatorStatus(ctx context.Context, resourc
...
+		FieldManager: fieldManager,
+	})
+	if err != nil {
+		return fmt.Errorf("unable to Apply Status for network operator using fieldManager %q: %w", fieldManager, err)
+	}
+
+	return nil
+}
+
+func (c *OperatorHelperClient) PatchOperatorStatus(ctx context.Context, jsonPatch *jsonpatch.PatchSet) (err error) {
```
[the remark this answers is not in the document]
Comment (jluhrsen) [reply pkg/client/operatorclient.go:136]: @jcaamano @p0lyn0mial , I've implemented it using [this](https://github.com/openshift/library-go/pull/1821/files#diff-3e6eecdb37386d21a87863b978e3dcf05ef2a0ecce8efd37822a21d02bfb02e4) as a reference. Look ok?

[the remark this answers is not in the document]
Comment (atiratree) [reply pkg/client/operatorclient.go:136]: The rationale behind this was that we should strive to stop using patch operations.

**Pull request merged**